### PR TITLE
Openblt

### PR DESCRIPTION
**Pull request by dron0gus**
This adds OpenBLT bootloader and tools as submodule to `firmware/ext/`
Also OpenBLT is build if `USE_OPENBLT` set to yes before calling `common_make.sh`
Boards specific OpenBLT code is expected at `firmware/config/boards/<board_name>/openblt/`. See OpenBLT `Target/Demo` or RusEFI's `boards/subaru_eg33` for example.
Initially OpenBLT can be flashed using DFU tool.
To flash main FW DFU util can be used.
Currently CRC check of main application is disabled in OpenBLT (for subaru board) and it is possible to load main application using DFU.
```
$ time ./BootCommander -t=xcp_can -d=can0 /home/work/rusefi/rusefi/firmware/build/rusefi.srec 
--------------------------------------------------------------------------
BootCommander version 1.04.02. Performs firmware updates on a micro-
controller based system that runs the OpenBLT bootloader.

Copyright (c) 2017 by Feaser  http://www.feaser.com
-------------------------------------------------------------------------
Processing command line parameters...[OK]
Detected firmware file: /home/work/rusefi/rusefi/firmware/build/rusefi.srec
Detected session protocol: XCP version 1.0
Using session protocol settings:
  -> Timeout T1: 1000 ms
  -> Timeout T3: 2000 ms
  -> Timeout T4: 10000 ms
  -> Timeout T5: 1000 ms
  -> Timeout T6: 50 ms
  -> Timeout T7: 2000 ms
  -> Seed/Key file: None
  -> Connection mode: 0
Detected transport layer: XCP on CAN
Using transport layer settings:
  -> Device: can0 (channel 0)
  -> Baudrate: 500000 bit/sec
  -> Transmit CAN identifier: 667h
  -> Receive CAN identifier: 7E1h
  -> Use 29-bit CAN identifiers: No
Loading firmware data from file...[OK]
  -> Number of segments: 3
  -> Base memory address: 0x08008000
  -> Total data size: 520908 bytes
Connecting to target bootloader...[OK]
Erasing 520 bytes starting at 0x08008000...[OK]  
Erasing 424732 bytes starting at 0x08008210...[OK]  
Erasing 95656 bytes starting at 0x0806fd30...[OK]  
Programming 520 bytes starting at 0x08008000...[OK]  
Programming 424732 bytes starting at 0x08008210...[OK]  
Programming 95656 bytes starting at 0x0806fd30...[OK]  
Finishing programming session...[OK]

real	1m35,129s
user	1m37,810s
```

### CODE REVIEW

**Comment by rusefillc** [comment]
PR says 80 files added. Does not look right. Submodule not added right?

**Comment by dron0gus** [comment]
> PR says 80 files added. Does not look right. Submodule not added right?
This is how OpenBLT sources are organized. Each target contains it's own copy of STM32 (or other MCU) HAL and CMSIS.
Another option is to keep RusEFI targets in our fork of OpenBLT.

File: firmware/config/boards/subaru_eg33/openblt/led.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/************************************************************************************//**
+* \file         Demo/ARMCM7_STM32F7_Nucleo_F767ZI_GCC/Boot/led.c
+* \brief        LED driver source file.
```
Comment (rusefillc) [inline firmware/config/boards/subaru_eg33/openblt/led.c:3]: please move out of boards/ into hw_layer/stm32/f7 for other boards to reuse
please PR all openblt files in a separate PR and integration with rusEFI as another one

File: firmware/hw_layer/ports/stm32/stm32f7/STM32F7.ld [hunk 2]
```diff
@@ -27,8 +27,8 @@
 MEMORY
 {
     bl          : org = 0x08000000, len = 16k       /* bootloader section */
-    flash0 (rx) : org = DEFINED(BOOTLOADER) ? 0x08008000 : 0x08000000, len = DEFINED(BOOTLOADER) ? 736k : 768k  /* change address & length if bootloader */
-    flash1 (rx) : org = 0x00200000, len = 2M        /* Flash as ITCM */ /* TODO: do we need BOOTLOADER logic here? */
+    flash0 (rx) : org = DEFINED(BOOTLOADER) ? 0x08008000 : 0x08000000, len = DEFINED(BOOTLOADER) ? (768k - 32k) : 768k  /* change address & length if bootloader */
```
Comment (rusefillc) [inline firmware/hw_layer/ports/stm32/stm32f7/STM32F7.ld:30]: is this an unrelated fix? please PR separately

File: firmware/config/boards/common_make.sh [hunk 3]
```diff
@@ -25,6 +25,23 @@ fi
...
+  [ -e build/openblt_$PROJECT_BOARD.hex ] || { echo "FAILED to compile OpneBLT by $SCRIPT_NAME with $PROJECT_BOARD"; exit 1; }
+  rm -f deliver/openblt_$PROJECT_BOARD.dfu
+  echo "$SCRIPT_NAME: invoking hex2dfu for OpenBLT"
+  if uname | grep "NT"; then
+   chmod u+x ../misc/encedo_hex2dfu/hex2dfu.exe
+   ../misc/encedo_hex2dfu/hex2dfu.exe -i build/openblt_$PROJECT_BOARD.hex -o deliver/openblt_$PROJECT_BOARD.dfu
+  else
+   chmod u+x ../misc/encedo_hex2dfu/hex2dfu.bin
+   ../misc/encedo_hex2dfu/hex2dfu.bin -i build/openblt_$PROJECT_BOARD.hex -o deliver/openblt_$PROJECT_BOARD.dfu
+  fi
```
Comment (rusefillc) [inline firmware/config/boards/common_make.sh:40]: can we have combined .dfu file so that my process does not change?

wait, would that mean two different DFU files - one for USB DFU (bl+rusefi) and another one for CAN update (only re?)

File: firmware/Makefile [hunk 4]
```diff
@@ -387,5 +400,11 @@ ULIBS = -lm --specs=nano.specs
 
 include $(RULESFILE)
 
+openblt:
+	make -C $(PROJECT_DIR)/config/boards/$(PROJECT_BOARD)/openblt/
```
Comment (rusefillc) [inline firmware/Makefile:404]: i am in favor of bootloader pre-combilation mainly because I compile rusEFI too often. on the bright side I would not be affected as long as we do not enable bootloader into default 407 Frankenso bundle

**Review by rusefillc — CHANGES_REQUESTED** [review]

**Comment by rusefillc** [comment]
Some technical notes at https://github.com/rusefi/rusefi/wiki/Firmware-update-via-CAN

File: firmware/config/boards/common_make.sh (hunk 3 above)
Comment (dron0gus) [reply firmware/config/boards/common_make.sh:40]: Yes. Two: One for (initial) flashing of board (OpenBLT+RusEFI combined). And second for RusEFI update: dfu or srec, dependent on method of update.
OpenBLT tools expect *.srec files, not dfu. But we can learn it to understand dfu too.

File: firmware/hw_layer/ports/stm32/stm32f7/STM32F7.ld (hunk 2 above)
Comment (dron0gus) [reply firmware/hw_layer/ports/stm32/stm32f7/STM32F7.ld:30]: Ok

File: firmware/Makefile (hunk 4 above)
Comment (dron0gus) [reply firmware/Makefile:404]: Build on request and keep prebuild images in git? Same strategy as for autogenerated.

Comment (rusefillc) [reply firmware/Makefile:404]: I do not insist on prebuild images in git, that's something we can decide on later. just a note that compilation time matters - specifically default build compilation time

Comment (mck1117) [reply firmware/Makefile:404]: I am strongly against prebuilt images in git - we don't have to compile this for every board, only ones that actually need it.

Comment (rusefillc) [reply firmware/Makefile:404]: @mck1117 but the cheapest maintenance is to have it on ALL builds, and having in ALL builds would mean longer compile

Comment (mck1117) [reply firmware/Makefile:404]: This change, as it sits, only compiles it when we need it.

Comment (rusefillc) [reply firmware/Makefile:404]: For the next 60 days we absolutely do NOT need prebuilt images. Let's move forward with everything else for now.

File: firmware/config/boards/subaru_eg33/openblt/led.c (hunk 1 above)
Comment (dron0gus) [reply firmware/config/boards/subaru_eg33/openblt/led.c:3]: Compromise solution: reuse HAL and CMSIS files from OpenBLT Demo project. As we do not heed this files for RE itself there is no reason to have it RE git.
Only two additional files have to be added for HAL configuration: config header and system.c file with CPU configuration.

**Comment by dron0gus** [comment]
Added support for MRE.

CRC checking is still disabled.
OpenBLT stores magic number at magic position in user FW (unused IRQ vector). And check this magic at boot. While image flashed over DFU does not have this magic numbers.
While DFU method is preferred CRC check is disabled. This cause OpenBLT to jump to empty flash on blank devices.

**Comment by rusefillc** [comment]
Is it time to create a ticket with OpenBLT or to start patching OpenBLT to remove that validation more competely?

**Comment by mck1117** [comment]
why do we not put a CRC in that spot in our vector table?

**Comment by rusefillc** [comment]
@mck1117 because level of effort unless you have hints for how to make this happen with little effort?

**Comment by dron0gus** [comment]
> Is it time to create a ticket with OpenBLT or to start patching OpenBLT to remove that validation more completely?

No problem to disable it through blt_conf.h - as it is done now . Someday we will need it.

File: firmware/config/boards/microrusefi/openblt/STM32F429ZITx_FLASH.ld [hunk 5]
```diff
@@ -0,0 +1,222 @@
+/*
```
Comment (rusefillc) [inline firmware/config/boards/microrusefi/openblt/STM32F429ZITx_FLASH.ld:1]: .ld files are not board specific but MCU specific right? let's move out of MRE folder

**Review by rusefillc — CHANGES_REQUESTED** [review]

File: firmware/config/boards/microrusefi/openblt/STM32F429ZITx_FLASH.ld (hunk 5 above)
Comment (dron0gus) [reply firmware/config/boards/microrusefi/openblt/STM32F429ZITx_FLASH.ld]: Done.

File: firmware/config/boards/microrusefi/compile_mre_f4.sh [hunk 6]
```diff
@@ -6,4 +6,6 @@ export EXTRA_PARAMS="-DSHORT_BOARD_NAME=mre_f4"
 
 # export DEBUG_LEVEL_OPT="-O0 -ggdb -g"
 
+export USE_OPENBLT=yes
```
Comment (rusefillc) [inline firmware/config/boards/microrusefi/compile_mre_f4.sh:9]: let's add same into Proteus F4 and Proteus F7 just to leverage HW CI

not sure which version we actually use for HW CI @mck1117 ?

**Review by rusefillc — CHANGES_REQUESTED** [review]

File: firmware/config/boards/microrusefi/compile_mre_f4.sh (hunk 6 above)
Comment (dron0gus) [reply firmware/config/boards/microrusefi/compile_mre_f4.sh:9]: Done for F4 and F7.
